### PR TITLE
Allow to set throw quality and spin odds

### PR DESCRIPTION
**Pull request by Likael**
This is a new PR that replaces #2286

**_WARNING:_**
The  code below is safe to use.
However, I tried to test further the possibilities of the field _normalized_hit_position_, and, coincidence or not, couldn't log again with my account any more for 1 hour immediately after sending a weird value. If you want to proceed and modify _normalized_hit_position_ to test randomization, be careful.

**Short Description:** 
The code allows to modify the frequency of the various throw possibilities.
For example, with the following setting : 
"catch_throw_parameters": {
	  "excellent_rate": 4,
	  "great_rate": 10,
	  "nice_rate": 4,
	  "normal_rate": 2,
	  "spin_success_rate" : 0.6,
	}
The bot would have 4 chances out of  20 (4+10+4+2) to land an _Excellent_ throw.
In the sample config, those value add up to 1 to look like percentage, but the user is free to set the values to what speaks most to him/her.
The _spin_success_rate_ is a percentage of landing a successful spin. In the example above, the bot has 60% chance of landing a spin that will generate the bonus exp. 

**More information about the _normalized_hit_position_ :** 
The code below should work safely as it doesn't contain the further tests i have tried, however my paranoid nature would ask for people to test with account they don't care about for 1 day before merging it. I will do so myself tomorrow too. Since the (unban?) I have been testing the code below, and got no further problem.
The value of 0 for _normalized_hit_position_  is safe, as it was what we were sending before I correct the API call.
The value of 1.0 is also safe, as we used it forever.
I didn't test any values between 0 et 1.
Values above 1 seems to be what could have caused the potential temporary ban.
The last case don't happen with this code, as the value is set either to 0 or 1.

**Fixes:**
- #2250

### CODE REVIEW

**Comment by maryokhin** [comment]
According to this [Reddit post](https://www.reddit.com/r/pokemongodev/comments/4w1cvr/pokemongo_current_api_status/), something like this will potentially flag all accounts using the bot.
> That said we would like to confirm that Niantic can detect any MITM apps, these are apps that somehow modify data sent to the server. For example an app that ensures a perfect pokeballthrow. If you used an app like that Niantic could know.

**Comment by maryokhin** [comment]
ah, ok, I thought it was the other way around. IMO the bot should just have a default randomness on how it throws the ball and have a setting to "enable perfect throws". After all, one of the goals of a bot is to emulate a human.

**Comment by Likael** [comment]
Well, the problem with default randomness is that it is only pseudo-random. Without inputs, you take the risk to see all bots behaving the same on the long term. By allowing people to modify inputs, you generate a greater variety of behaviours, which makes each bot more unique. Of course some combinations could and/or would trigger flag (i.e. 100% perfect throw, or any behaviours that move too far away from the standard deviation), but it is not ours to judge. The exercise here is to challenge ourselves to design a bot that is able to get as close as possible of a human behaviour.

**Comment by douglascamata** [comment]
Need to solve merge conflicts.

**Comment by RedViper9** [comment]
@Likael Would it be possible to also add a "miss" option for the sake of more human behavior?

**Comment by Likael** [comment]
@douglascamata Done.
@RedViper9 Sure, but I first need to have a check at the missed throw requests to ensure the data we will be sending are coherent with the app one.

File: pokemongo_bot/cell_workers/pokemon_catch_worker.py [hunk 1]
```diff
@@ -432,3 +441,50 @@ def _do_evolve(self, pokemon, new_pokemon_id):
                 formatted='Failed to evolve {pokemon}!',
                 data={'pokemon': pokemon.name}
             )
+
+    def generate_random_throw_parameters_if_necessary(self, throw_parameters):
```
Comment (douglascamata) [inline pokemongo_bot/cell_workers/pokemon_catch_worker.py]: can you please break down this method into smaller ones?

**Comment by douglascamata** [comment]
@Likael see my comment and fix merge conflicts again, please? :D

**Comment by Likael** [comment]
@douglascamata Partly done. I merged, but couldn't really break down the code. I have extracted the spin function. The rest is as it is, just the randomization part. I unfolded it for clarity and easy debug, but it's possible de condensate it using a for loop and creating the adequate data structure in the class. However it will lose the clarity. I don't really have time to do that today, but if that's what you want, I can have a look this weekend.

**Comment by douglascamata** [comment]
@Likael add an example configuration to teach us how to configure these parameters, please.

**Comment by Likael** [comment]
@douglascamata Sorry the example has been trimmed away in the previous commit. Added it again.

**Comment by douglascamata** [comment]
:+1: 

[Approved with PullApprove](https://pullapprove.com/PokemonGoF/PokemonGo-Bot/pull-request/2534/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=PokemonGoF/PokemonGo-Bot)

**Comment by patrickbrouwers** [comment]
The example config is invalid: Strict JSON does not allow a final comma in an object (dictionary) literal (https://github.com/PokemonGoF/PokemonGo-Bot/pull/2534/commits/ec82e9475b6f4259abcdc6cd33bb7764c7788108#diff-12669c2a3cbb1440601e0581913cff68R109)

**Comment by douglascamata** [comment]
@patrickbrouwers thx, already pushed a fix for this a few minutes ago.

**Pull request merged**